cmake/genny_subdir.cmake

```diff
@@ -0,0 +1,223 @@
+# -   Assumes source files are in `src` and end with `.cpp`.
+# -   The file `src/main.cpp` is special. It is not compiled into the
+#     produced library. It is only compiled if the `EXECUTABLE` argument
+#     is given.
+# -   Assumes regular tests are `test/*.cpp` and get parsed by
+#     `ParseAndAddCatchTests`. The test binary is named
+#     `project_name_test`.
+# -   Assumes benchmark tests are `benchmark/*.cpp` and get parsed by
+#     `ParseAndAddCatchTests`. The benchmark test binary is named
+#     `project_name_benchmark`.
+# -   Installs the library to the `GennyLibraryConfig` export
+#
+function(GENNY_SUBDIR)
+    set(_gs_name) # value from NAME
+    set(_gs_want_name 0) # prev arg was NAME
+
+    set(_gs_type) # value from TYPE
+    set(_gs_want_type 0) # prev arg was TYPE
+
+    set(_gs_executable) # value from EXECUTABLE
+    set(_gs_want_executable 0) # prev arg was EXECUTABLE
+
+    set(_gs_depends) # list for DEPENDS
+    set(_gs_test_depends) # list for TEST_DEPENDS
+    set(_gs_list_target "") # if we're in a list, which list?
+
+    ## Parse args to set _gs_type etc
+
+    # for each arg we look for caps keywords
+    # e.g. NAME and set the appropriate _gs_want_X
+    # var. Then on next iteration if we have the
+    # _gs_want_X we set the _gs_X to the current value
+    # and unset the _gs_want var. For list types
+    # e.g. DEPENDS/TEST_DEPENDS we append to the
+    # appropriate lists.
+    foreach(arg IN LISTS ARGV)
+        # we could be more pedantic in here
+        # to assert that we have at most
+        # one _gs_want* set but that seems..icky.
+        if(arg MATCHES "^NAME$")
+            set(_gs_want_name 1)
+            continue()
+        elseif(_gs_want_name)
+            set(_gs_name "${arg}")
+            set(_gs_want_name 0)
+            continue()
+        elseif(arg MATCHES "^TYPE$")
+            set(_gs_want_type 1)
+            continue()
+        elseif(_gs_want_type)
+            set(_gs_type "${arg}")
+            set(_gs_want_type 0)
+            continue()
+        elseif(arg MATCHES "^EXECUTABLE")
+            set(_gs_want_executable 1)
+            continue()
+        elseif(_gs_want_executable)
+            set(_gs_executable "${arg}")
+            set(_gs_want_executable 0)
+            continue()
+        elseif(arg MATCHES "^DEPENDS$")
+            # append values (see else below) to _gs_depends
+            set(_gs_list_target "_gs_depends")
+            continue()
+        elseif(arg MATCHES "^TEST_DEPENDS$")
+            # append values (see else below) to _gs_test_depends
+            set(_gs_list_target "_gs_test_depends")
+            continue()
+        else()
+            # if no _gs_want* set and no keyword types
+            # like NAME/TYPE etc then we must be in a
+            # list i.e. DEPENDS or TEST_DEPENDS so
+            # add the arg to teh appropriate list
+            list(APPEND "${_gs_list_target}" "${arg}")
+        endif()
+    endforeach()
```

Comment:
I decided against being super pedantic about parsing errors e.g. handling missing name/type value or specifying e.g. `genny_subdir(TYPE NAME foo)` or whatever...I don't think we'll add new subdirs that often and the errors that come from improper usage were (to me) pretty obvious when I was playing around with it.

Reply:
Noted. Have you looked into `https://cmake.org/cmake/help/latest/command/cmake_parse_arguments.html`? It seems to provide most of the functionality you need but I'm not sure how well it does error handling?

Reply:
Yeah, just use [cmake_parse_arguments](https://cmake.org/cmake/help/latest/command/cmake_parse_arguments.html#command:cmake_parse_arguments).

Reply:
Done